fmp_server.py

```diff
@@ -305,18 +305,51 @@ def recvReqLayerUp(self, packet_id, data, seq):
          - Relocate > Select Server
         """
         if self.session.layer == 2:
-            city = self.session.get_city()
-            if city.leader == self.session:
-                if len(city.players) > 1:
-                    # TODO: Transfer the leadership to another player
-                    pass
-                city.leader = None
+            new_leader = self.session.try_transfer_city_leadership()
+            if new_leader is not None:
+                self.sendNtcLayerHost(new_leader, seq)
 
             ntc_data = pati.lp2_string(self.session.capcom_id)
             self.server.layer_broadcast(self.session, PatID4.NtcLayerOut,
                                         ntc_data, seq)
         self.sendAnsLayerUp(data, seq)
 
+    def recvReqLayerEnd(self, packet_id, data, seq):
+        """ReqLayerEnd packet.
+
+        ID: 64020100
+        JP: レイヤ終了要求
+        TR: Layer end request
+        """
+        if self.session.layer > 1:
+            # City path
+            new_leader = self.session.try_transfer_city_leadership()
+            if new_leader is not None:
+                self.sendNtcLayerHost(new_leader, seq)
+
+            ntc_data = pati.lp2_string(self.session.capcom_id)
+            self.server.layer_broadcast(self.session, PatID4.NtcLayerOut,
+                                        ntc_data, seq)
+            self.session.leave_city()
+        if self.session.layer > 0:
+            # Gate path
+            self.session.leave_gate()
+            # Server path (executed at gate and higher)
+            self.session.leave_server()
+
+        self.sendAnsLayerEnd(seq)
+    
+    def sendNtcLayerHost(self, new_leader, seq):
+        """NtcLayerHost packet.
+        ID: 64411000
```

Comment:
It's missing a newline here.

Reply:
Only *just* realized what was missing the newline here. It will be fixed in the next commit.